internal/controllers/manager.go

```diff
@@ -6,45 +6,54 @@ import (
 	"fmt"
 
 	"k8s.io/apimachinery/pkg/runtime"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/client-go/rest"
 	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
 	"github.com/tilt-dev/tilt/internal/hud/server"
 	"github.com/tilt-dev/tilt/internal/store"
-	corev1alpha1 "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
 	"github.com/tilt-dev/tilt/pkg/logger"
 )
 
 type TiltServerControllerManager struct {
-	config *rest.Config
-	cancel context.CancelFunc
+	config  *rest.Config
+	scheme  *runtime.Scheme
+	manager ctrl.Manager
+	cancel  context.CancelFunc
 }
 
 var _ store.SetUpper = &TiltServerControllerManager{}
 var _ store.Subscriber = &TiltServerControllerManager{}
 var _ store.TearDowner = &TiltServerControllerManager{}
 
-func NewTiltServerControllerManager(config *server.APIServerConfig) *TiltServerControllerManager {
+func NewTiltServerControllerManager(config *server.APIServerConfig, scheme *runtime.Scheme) (*TiltServerControllerManager, error) {
 	return &TiltServerControllerManager{
 		config: config.GenericConfig.LoopbackClientConfig,
+		scheme: scheme,
+	}, nil
+}
+
+func (m *TiltServerControllerManager) GetManager() ctrl.Manager {
+	return m.manager
+}
+
+func (m *TiltServerControllerManager) GetClient() ctrlclient.Client {
+	if m.manager == nil {
+		return nil
 	}
+	return m.manager.GetClient()
 }
 
 func (m *TiltServerControllerManager) SetUp(ctx context.Context, st store.RStore) error {
-	scheme := runtime.NewScheme()
-
 	ctx, m.cancel = context.WithCancel(ctx)
 
 	// TODO(milas): we should provide a logr.Logger facade for our logger rather than using zap
 	w := logger.Get(ctx).Writer(logger.DebugLvl)
 	ctrl.SetLogger(zap.New(zap.UseDevMode(true), zap.WriteTo(w)))
 
-	utilruntime.Must(corev1alpha1.AddToScheme(scheme))
-
 	mgr, err := ctrl.NewManager(m.config, ctrl.Options{
```

Comment:
can we dependency-inject the manager? 

i think that would avoid a lot of the circular initialization issues you're running into

Reply:
Take a look at cc45e52d5 and see if you think that's an improvement? It really just lets controllers pretend to get created as expected by getting a `ctrlclient.Client` injected. The impl is `*deferredClient` that returns errors (or `nil` for methods that don't return `error` which is less than ideal) until the manager initializes it.

Not sure that the extra indirection is worth saving on needing to call `SetClient()`, but at least it hides some of the oddities from each controller implementation.